server/controllers/auth.controller.js

```diff
@@ -44,17 +45,86 @@ function login(req, res, next) {
             scopes: userResult.scopes,
         };
 
-        const token = signJWT(userInfo);
+        const jwtToken = signJWT(userInfo);
+        const refreshToken = randtoken.uid(128);
+
+        // save the refresh token
+        userResult.update({ refreshToken });
 
         return res.json({
-            token,
+            token: jwtToken,
             username: user.username,
+            refreshToken,
         });
     })
     .catch(error => next(error));
 }
 
-function logout() {}
+/**
+ * Sends back jwt token if valid username and password is provided
+ * @param req
+ * @param res
+ * @param next
+ * @returns {*}
+ */
+function submitRefreshToken(req, res, next) {
+    const params = _.pick(req.body, 'username', 'refreshToken');
+    User.findOne({ where: {
+        username: params.username,
+        refreshToken: params.refreshToken,
+    } })
+    .then((userResult) => {
+        if (_.isNull(userResult)) {
+            const err = new APIError('Refresh token not found', httpStatus.NOT_FOUND, true);
+            return next(err);
+        }
+        const userInfo = {
+            id: userResult.id,
+            username: userResult.username,
+            email: userResult.email,
+            scopes: userResult.scopes,
+        };
+
+        const jwtToken = signJWT(userInfo);
+        const refreshToken = randtoken.uid(128);
```

Comment:
It's almost no extra work for the client, but is there a reason to generate a new refresh token here in addition to the access token? I haven't seen that done in my reading.

Reply:
I think the idea is that a refresh token should be single-use.